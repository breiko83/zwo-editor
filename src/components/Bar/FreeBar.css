.segment {
  position: relative;
}
.freeRide {    
  border-radius: 0px;
  border: 1px solid white;
  background-image: url('../../assets/freeride.svg');
}