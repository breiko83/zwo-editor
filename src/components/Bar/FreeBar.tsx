import React, { useState } from 'react'
import './FreeBar.css'
import { Resizable } from 're-resizable'
import 'moment-duration-format'
import Label from '../Label/Label'
import { FreeInterval } from '../../types/Interval'
import { durationMultiplier } from './multipliers'
import { WorkoutMode } from '../../modes/WorkoutMode'

interface FreeBarProps {
  interval: FreeInterval;
  mode: WorkoutMode;
  onChange: (interval: FreeInterval) => void;
  onClick: (id: string) => void;
  selected: boolean;
}

const FreeBar = ({interval, mode, ...props}: FreeBarProps) => {
  const [width, setWidth] = useState(mode.lengthToWidth(interval.length))

  const [showLabel, setShowLabel] = useState(false)

  const handleCadenceChange = (cadence: number) => {
    props.onChange({ ...interval, cadence })
  }

  // standard height
  const height = 100

  const handleResizeStop = (dWidth: number) => {
    setWidth(width + dWidth)
    notifyChange(dWidth)
  }

  const notifyChange = (dWidth: number) => {
    props.onChange({ ...interval, length: mode.widthToLength(width + dWidth) })
  }

  return (
    <div className='segment'
      onMouseEnter={() => setShowLabel(true)}
      onMouseLeave={() => setShowLabel(false)}
      style={props.selected ? {zIndex:1}: {}}
      onClick={() => props.onClick(interval.id)}
    >
      {(props.selected || showLabel) &&
        <Label
          interval={interval}
          mode={mode}
          onCadenceChange={(cadence: number)=> handleCadenceChange(cadence)}
        />
      }
      <Resizable
        className='freeRide'
        size={{
          width: mode.lengthToWidth(interval.length),
          height: height,
        }}
        minWidth={durationMultiplier}
        minHeight={height}
        maxHeight={height}
        enable={{ right: true }}
        grid={[1, 1]}
        onResizeStop={(e, direction, ref, d) => handleResizeStop(d.width)}
        onResize={(e, direction, ref, d) => notifyChange(d.width)}        
      >
      </Resizable>
    </div>
  )

}

export default FreeBar
