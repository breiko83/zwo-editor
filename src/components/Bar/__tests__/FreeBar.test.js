import React from 'react';
import FreeBar from '../FreeBar';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect'
import { v4 as uuidv4 } from 'uuid'
import { Zones } from '../../Constants'

test('FreeBar renders correctly', () => {
  const interval = {
    time: 50,
    power: Zones.Z3.min,
    type: 'steady',
    id: uuidv4()
  }

  const component = renderer.create(
    <FreeBar
      interval={interval}
      onChange={() => { }}
      onClick={() => { }}
    />
  )

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})