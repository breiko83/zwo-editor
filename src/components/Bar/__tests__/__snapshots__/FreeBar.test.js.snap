// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FreeBar renders correctly 1`] = `
<div
  className="segment"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={Object {}}
>
  <div
    className="freeRide"
    style={
      Object {
        "boxSizing": "border-box",
        "flexShrink": 0,
        "height": "100px",
        "maxHeight": 100,
        "maxWidth": undefined,
        "minHeight": 100,
        "minWidth": 3,
        "position": "relative",
        "userSelect": "auto",
        "width": "16.666666666666668px",
      }
    }
  >
    <span>
      <div
        className=""
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "cursor": "col-resize",
            "height": "100%",
            "position": "absolute",
            "right": "-5px",
            "top": "0px",
            "userSelect": "none",
            "width": "10px",
          }
        }
      />
    </span>
  </div>
</div>
`;
